core/modules/widgets/widget.js

```diff
@@ -596,10 +596,28 @@ Widget.prototype.findFirstDomNode = function() {
 	return null;
 };
 
+
+
+/*
+Depth first destroy
+*/
+Widget.prototype.desendDestroy = function() {
+	this.waitDestroy = true; //for blocking repeat calls to this method from removeChildDomNodes
+	$tw.utils.each(this.children,function(childWidget) {
+		childWidget.desendDestroy();
+	});
+	
+	if (this.destroy) {
+		this.destroy();
+	}
+	this.waitDestroy = false;
+};
+
+Widget.prototype.waitDestroy = false;
 /*
 Remove any DOM nodes created by this widget or its children
 */
-Widget.prototype.removeChildDomNodes = function() {
+Widget.prototype.removeChildDomNodes = function(noDestroy) {
```

Comment:
what would be the opposite of "noDestroy" .... I think I do have a problem with the "inverted" logic in combination with the inverted variable names.

Reply:
All widgets call removeChildDomNodes() with no parameter, which is falsy, and I cannot pass in 'false'

Reply:
Yea, I think I know how it works, but it needs a bit of mind-twisting to get it. So if there would be "easier" names it may help

Reply:
> All widgets call removeChildDomNodes() with no parameter, which is falsy, and I cannot pass in 'false'

Perhaps we could check explicitly for undefined, and not just treat it as falsy?

Reply:
we could pass in an options object, like this:

```

Widget.prototype.removeChildDomNodes = function(options) {
	// If this widget has directly created DOM nodes, delete them and exit. This assumes that any child widgets are contained within the created DOM nodes, which would normally be the case
	if(this.domNodes.length > 0) {
		$tw.utils.each(this.domNodes,function(domNode) {
			domNode.parentNode.removeChild(domNode);
		});
		this.domNodes = [];
	} else {
		// Otherwise, ask the child widgets to delete their DOM nodes
		$tw.utils.each(this.children,function(childWidget) {
			childWidget.removeChildDomNodes({destroy:true}); 
		});
	}
	if ((options && options.destroy && !this.waitDestroy) this.desendDestroy();
};
```